src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -2276,124 +2276,139 @@ private MemberDeclarationSyntax ParseMemberDeclarationOrStatementCore(SyntaxKind
-                        return null;
+                        if (attributes.Count == 0 && modifiers.Count == 0 && type.IsMissing && !sawRef)
+                        {
+                            // we haven't advanced, the caller needs to consume the tokens ahead
+                            return null;
+                        }
+
+                        var incompleteMember = _syntaxFactory.IncompleteMember(attributes, modifiers.ToList(), type.IsMissing ? null : type);
+                        if (incompleteMember.ContainsDiagnostics)
+                        {
+                            return incompleteMember;
+                        }
+                        else if (parentKind == SyntaxKind.NamespaceDeclaration ||
+                                 parentKind == SyntaxKind.CompilationUnit && !IsScript)
+                        {
+                            return this.AddErrorToLastToken(incompleteMember, ErrorCode.ERR_NamespaceUnexpected);
+                        }
+                        else
+                        {
+                            //the error position should indicate CurrentToken
+                            return this.AddError(
+                                incompleteMember,
+                                incompleteMember.FullWidth + this.CurrentToken.GetLeadingTriviaWidth(),
+                                this.CurrentToken.Width,
+                                ErrorCode.ERR_InvalidMemberDecl,
+                                this.CurrentToken.Text);
+                        }
                     }
 
-                    var incompleteMember = _syntaxFactory.IncompleteMember(attributes, modifiers.ToList(), type.IsMissing ? null : type);
-                    if (incompleteMember.ContainsDiagnostics)
+                    // If the modifiers did not include "async", and the type we got was "async", and there was an
+                    // error in the identifier or its type parameters, then the user is probably in the midst of typing
+                    // an async method.  In that case we reconsider "async" to be a modifier, and treat the identifier
+                    // (with the type parameters) as the type (with type arguments).  Then we go back to looking for
+                    // the member name again.
+                    // For example, if we get
+                    //     async Task<
+                    // then we want async to be a modifier and Task<MISSING> to be a type.
+                    if (!sawRef &&
+                        identifierOrThisOpt != null &&
+                        (typeParameterListOpt != null && typeParameterListOpt.ContainsDiagnostics
+                          || this.CurrentToken.Kind != SyntaxKind.OpenParenToken && this.CurrentToken.Kind != SyntaxKind.OpenBraceToken && this.CurrentToken.Kind != SyntaxKind.EqualsGreaterThanToken) &&
+                        ReconsiderTypeAsAsyncModifier(ref modifiers, type, identifierOrThisOpt))
```

Comment:
this if-statement is hte poster child for something that breaks my brain.  i literally can't understand it.

Reply:
this also seems strange in that i would have expected ParseModifiers to have this logic.  When we see 'async' i would think we could make the decision there (using whatever lookahead is necessary) to think it's a mod or not.  instead of needing to undo our original decision back there.

Reply:
@AlekseyTs, do you have any thoughts on this?

---
In reply to: [252526047](https://github.com/dotnet/roslyn/pull/32983#discussion_r252526047) [](ancestors = 252526047)

Reply:
I prefer us to not do any refactoring under umbrella of this fix. It will make reviewing and proving correctness of the fix much harder rather than easier. The goal of the fix is to adjust the way we reinterpret the sequence of tokens, we are not changing the conditions under which the sequence is reinterpreted.